src/detectors.rs

```diff
@@ -0,0 +1,340 @@
+    ///
+    fn blocking_detected(&self, workers: &[ThreadInfo]);
+}
+
+impl<F> BlockingActionHandler for F
+where
+    F: Fn(&[ThreadInfo]) + Send + Sync,
+{
+    fn blocking_detected(&self, workers: &[ThreadInfo]) {
+        // Implement the behavior for blocking_detected using the provided function.
+        // You can call the function here with the given workers.
+        self(workers);
+    }
+}
+struct StdErrBlockingActionHandler;
+
+/// BlockingActionHandler implementation that writes blocker details to standard error.
+impl BlockingActionHandler for StdErrBlockingActionHandler {
+    fn blocking_detected(&self, workers: &[ThreadInfo]) {
+        eprintln!("Detected blocking in worker threads: {:?}", workers);
+    }
+}
+
+#[derive(Debug)]
+struct WorkerSet {
+    inner: Mutex<HashSet<ThreadInfo>>,
+}
+
+impl WorkerSet {
+    fn new() -> Self {
+        WorkerSet {
+            inner: Mutex::new(HashSet::new()),
+        }
+    }
+
+    fn add(&self, pid: ThreadInfo) {
+        let mut set = self.inner.lock().unwrap();
+        set.insert(pid);
+    }
+
+    fn remove(&self, pid: ThreadInfo) {
+        let mut set = self.inner.lock().unwrap();
+        set.remove(&pid);
+    }
+
+    fn get_all(&self) -> Vec<ThreadInfo> {
+        let set = self.inner.lock().unwrap();
+        set.iter().cloned().collect()
+    }
+}
+
+/// Utility to help with "really nice to add a warning for tasks that might be blocking"
+#[derive(Debug)]
+pub struct LongRunningTaskDetector {
+    interval: Duration,
+    detection_time: Duration,
+    stop_flag: Arc<Mutex<bool>>,
+    workers: Arc<WorkerSet>,
+}
+
+async fn do_nothing(tx: mpsc::Sender<()>) {
+    // signal I am done
+    tx.send(()).unwrap();
+}
+
+fn probe(
+    tokio_runtime: &Arc<Runtime>,
+    detection_time: Duration,
+    workers: &Arc<WorkerSet>,
+    action: &Arc<dyn BlockingActionHandler>,
+) {
+    let (tx, rx) = mpsc::channel();
+    let _nothing_handle = tokio_runtime.spawn(do_nothing(tx));
+    let is_probe_success = match rx.recv_timeout(detection_time) {
+        Ok(_result) => true,
+        Err(_) => false,
+    };
+    if !is_probe_success {
+        let targets = workers.get_all();
+        action.blocking_detected(&targets);
+        rx.recv_timeout(get_panic_worker_block_duration()).unwrap();
+    }
+}
+
+/// Utility to help with detecting blocking in tokio workers.
+///
+/// # Example
+///
+///  ```
+///    use std::sync::Arc;
+///    use tokio_metrics::detectors::LongRunningTaskDetector;
+///
+///    let (lrtd, mut builder) = LongRunningTaskDetector::new_multi_threaded(
+///      std::time::Duration::from_millis(10),
+///      std::time::Duration::from_millis(100)
+///    );
+///    let runtime = builder.worker_threads(2).enable_all().build().unwrap();
+///    let arc_runtime = Arc::new(runtime);
+///    let arc_runtime2 = arc_runtime.clone();
+///    lrtd.start(arc_runtime);
+///    arc_runtime2.block_on(async {
+///     print!("my async code")
+///    });
```

Comment:
Please take out all of the extra spaces at the beginning of each line. Lines in documentation should have only one space between the text and the `///`.